ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/BlockTree.hs

```diff
@@ -223,3 +229,58 @@ prettyBlockTree blockTree =
       case AF.anchor frag of
         AF.AnchorGenesis     -> 0
         AF.Anchor slotNo _ _ -> slotNo + 1
+
+
+
+-- Constructs a map from each block in the tree to the (unique) AnchoredFragment
+-- from the anchor of the tree to that block.
+deforestBlockTree
+  :: forall blk. (HasHeader blk)
+  => BlockTree blk -> M.Map (HeaderHash blk) (AF.AnchoredFragment blk)
+deforestBlockTree (BlockTree trunk branches) =
+  let
+    -- Get the pairs (tip, fragment) for all the nonempty AnchoredFragments in fs,
+    -- and filter out any empty fragments.
+    getTips :: [AF.AnchoredFragment blk] -> [(HeaderHash blk, AF.AnchoredFragment blk)]
+    getTips fs = case fs of
+      [] -> []
+      fragment:rest -> case fragment of
+        AF.Empty _ -> getTips rest -- Should never happen!
+        _ AF.:> tip -> (blockHash tip, fragment) : getTips rest
```

Comment:
Can we write this without the explicit recursion? Seems like a perfect case for `foldMap`